### PR TITLE
multi gpu docs

### PR DESCRIPTION
**Pull request by ImmanuelSegol**
## Describe the changes

This PR...

## Linked Issues

Resolves #

### CODE REVIEW

**Review by jeremyfelder — APPROVED** [review]

File: docs/docs/icicle/multi-gpu.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+### GPU Server approach 
+
+This approach usually involves a single or multiple CPUs opening threads to read / write from multiple GPUs. You can think about it as a scaled up HOST - Device model.
+
+![alt text](image.png)
+
+This approach wont let us tackle larger computation sizes but it will allow us to compute multiple computations which we wouldn't be able to load onto a single GPU.
+
+For example lets say that you had to compute two MSMs of size 2^20 on a 16GB VRAM GPU you would normally have to perform them asynchronously. However, if you double the number of GPUs in your system you can now run them in parallel.
```
Comment (vhnatyk) [inline docs/docs/icicle/multi-gpu.md:26]: 2^20 will fit into 16GB RAM even quite more of these will, right?  even with precomputation - for example for BLS12-381 - 2^20 * (48+32) = 80MB

Comment (ImmanuelSegol) [reply docs/docs/icicle/multi-gpu.md:26]: Thanks! That's a typo will fix

File: docs/docs/icicle/multi-gpu.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+- [Rust Multi GPU APIs](./rust-bindings/multi-gpu.md)
+- C++ Multi GPU APIs
+
+
+## Best practices 
+
+- Never hardcode device IDs, if you want your software to take advantage of all GPUs on a machine use methods such as `get_device_count` to support arbitrary number of GPUs.
+
+- Launch one thread per GPU, to avoid nasty errors and hard to read code we suggest that for every GPU task you wish to launch you create a dedicated thread. This will make your code way more manageable, easy to read and performant.
```
Comment (vhnatyk) [inline docs/docs/icicle/multi-gpu.md:50]: umm - one CPU thread per GPU - actually you can do more tasks on that thread as long as they target the same GPU. Also the section imo needs a link  to https://developer.nvidia.com/blog/cuda-pro-tip-always-set-current-device-avoid-multithreading-bugs/

File: docs/docs/icicle/rust-bindings/multi-gpu.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+```rust
+match get_device() {
+    Ok(device_id) => println!("Current device ID: {}", device_id),
+    Err(e) => eprintln!("Failed to get current device: {:?}", e),
+}
+```
+
+## Device context API
+
+The `DeviceContext` is embedded into `NTTConfig`, `MSMConfig` and `PoseidonConfig`, meaning you can simple pass a `device_id` to your existing config an the same computation will be triggered on a different device automatically.
```
Comment (vhnatyk) [inline docs/docs/icicle/rust-bindings/multi-gpu.md:81]: _and_ typo?

File: docs/docs/icicle/multi-gpu.md (hunk 1 above)
Comment (DmytroTym) [inline docs/docs/icicle/multi-gpu.md:26]: I think size 2^20 MSM on bls12 curves should require less than 500 Mb. For bls12, 2^26 is probably the size when 1 MSM fits into 16 GB but 2 do not

File: docs/docs/icicle/multi-gpu.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+This approach involves a more sophisticated approach to multi GPU computation. Using technologies such as [GPUDirect, NCCL, NVSHMEM](https://www.nvidia.com/en-us/on-demand/session/gtcspring21-cwes1084/) and NVLink its possible to combine multiple GPUs and split a computation among different devices.
+
+This approach requires redesigning the algorithm at the software level to be compatible with splitting amongst devices. In some cases, to lower latency to a minimum, special inter GPU connections would be installed on a server to allow direct communication between multiple GPUs.
+
+
+# Writing ICICLE Code for Multi GPUs
+
+The approach we have taken for the moment is a GPU Server approach; we assume you have a machine with multiple GPUs and you wish to run some computation on each GPU.
+
+To dive deeper and learn about the API checkout the docs for our different ICICLE API
```
Comment (DmytroTym) [inline docs/docs/icicle/multi-gpu.md:40]: checkout -> check out

File: docs/docs/icicle/rust-bindings/multi-gpu.md (hunk 3 above)
Comment (vhnatyk) [reply docs/docs/icicle/rust-bindings/multi-gpu.md:81]: actually current implementation doesn't have  the "automatic"  - we just check device_id from config matches the current device id for the thread, so it won't be executed on wrong device

File: docs/docs/icicle/rust-bindings/multi-gpu.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+#### [`DeviceContext`](https://github.com/vhnatyk/icicle/blob/eef6876b037a6b0797464e7cdcf9c1ecfcf41808/wrappers/rust/icicle-cuda-runtime/src/device_context.rs#L11)
+
+Represents the configuration a CUDA device, encapsulating the device's stream, ID, and memory pool. The default device is always `0`, unless configured otherwise.
```
Comment (vhnatyk) [inline docs/docs/icicle/rust-bindings/multi-gpu.md:85]: _, unless configured otherwise_ probably should be removed -  I doubt it's possible

File: docs/docs/icicle/rust-bindings/multi-gpu.md (hunk 3 above)
Comment (DmytroTym) [inline docs/docs/icicle/rust-bindings/multi-gpu.md:81]: simple -> simply

Comment (DmytroTym) [reply docs/docs/icicle/rust-bindings/multi-gpu.md:81]: an -> and

File: docs/docs/icicle/rust-bindings/multi-gpu.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+##### Fields
+
+- **`stream: &'a CudaStream`**
+
+  A reference to a `CudaStream`. This stream is used for executing CUDA operations. By default, it points to a null stream CUDA's default execution stream.
+
+- **`device_id: usize`**
+
+  The index of the GPU currently in use. The default value is `0`, indicating the first GPU in the system.
```
Comment (vhnatyk) [inline docs/docs/icicle/rust-bindings/multi-gpu.md:103]: umm, assuming invocation command was prepended with  `CUDA_VISIBLE_DEVICES=2,3,7 ` in the system with 8 GPUs  - the `device_id` `0` will correspond to GPU with id 2, so technically a third GPU in the system

**Review by vhnatyk — CHANGES_REQUESTED** [review]
hi @ImmanuelSegol  - I see pr was merged  already 😊, looks great 👍🏻 and there are small notes to consider

**Comment by DmytroTym** [comment]
@vhnatyk @ImmanuelSegol maybe we can fix the issues in https://github.com/ingonyama-zk/icicle/pull/389

**Pull request merged**